test/test_nestedtensor.py

```diff
@@ -7818,131 +7822,539 @@ def __torch_dispatch__(self, func, types, args=..., kwargs=None):
+                and (
+                    (
+                        isinstance(sample.kwargs["dim"], int)
+                        and sample.kwargs["dim"] == sample.input._ragged_idx
+                    )
+                    or (
+                        isinstance(sample.kwargs["dim"], (tuple, list))
+                        and sample.input._ragged_idx in sample.kwargs["dim"]
+                    )
+                )
+            )
+        ),
+        name="nansum_unimplemented",
+    ),
+    # expected: reducing across the ragged dimension is not supported for non-contiguous
+    # nested tensors with holes
+    XFailRule(
+        error_type=RuntimeError,
+        error_msg=(
+            "reducing across the ragged dimension is not supported for non-contiguous "
+            "nested tensors with holes"
+        ),
+        match_fn=lambda device, dtype, op, sample: (
+            # min.reduction_with_dim and max.reduction_with_dim aren't associated with
+            # ReductionOpInfo entries sadly even though they're reductions
+            (isinstance(op, ReductionOpInfo) or "reduction_with_dim" in op.full_name)
+            and "noncontig_holes" in sample.name
+            and "dim" in sample.kwargs
+            and (
+                (
+                    isinstance(sample.kwargs["dim"], int)
+                    and sample.kwargs["dim"] == sample.input._ragged_idx
+                )
+                or (
+                    isinstance(sample.kwargs["dim"], (tuple, list))
+                    and sample.input._ragged_idx in sample.kwargs["dim"]
+                )
+            )
+        ),
+        name="ragged_dim_reduction_noncontig_holes",
+    ),
+    # expected: index_put() doesn't work on non-contiguous NJTs without ragged dimension indices
+    XFailRule(
+        error_type=RuntimeError,
+        error_msg="If ragged dimension is not part of indices, this only works on contiguous NJTs",
+        match_fn=lambda device, dtype, op, sample: (
+            op.full_name == "index_put"
+            and not sample.input.is_contiguous()
+            and len(sample.kwargs["indices"]) - 1 < sample.input._ragged_idx
+        ),
+        name="index_put_noncontig_holes_no_ragged_dim_indices",
+    ),
+    # expected: masked_select() doesn't work on non-contiguous NJTs
+    XFailRule(
+        error_type=ValueError,
+        error_msg="expected self to be a contiguous jagged layout NestedTensor",
+        match_fn=lambda device, dtype, op, sample: (
+            op.full_name == "masked_select" and not sample.input.is_contiguous()
+        ),
+        name="masked_select_noncontig",
+    ),
+    # expected: bmm / matmul sometimes use a to_padded_tensor() fallback which isn't
+    # supported for non-contig NJTs with holes
+    XFailRule(
+        error_type=RuntimeError,
+        error_msg="not supported for nested tensors with holes",
+        match_fn=lambda device, dtype, op, sample: (
+            op.full_name in {"bmm", "matmul"}
```

Comment:
I wonder if its better to have `ops={"bmm", "matmul"}` be a field on XFailRule, and call `match_fn` something like `sample_match_fn` instead.

With those two separate I can utilize the "xfail_db" to query xfails based on op and transposition/contiguity separately, e.g. what are all the xfails for this op, what are all the xfails that I can run into for non-contig transposed njts.

Reply:
I don't necessarily disagree with a hierarchy of xfails, but note that this specific approach does limit the flexibility some. For example, I regularly do things like check `isinstance(op, BinaryUfuncInfo)` which wouldn't be possible if we just had a list of op names.

To keep the existing flexibility: we could have both `op_fn` and `sample_fn`, where the former only needs to be called once per test and the latter needs to be called once per sample. This could help test runtime a tad (TBD how much).

> what are all the xfails that I can run into for non-contig transposed njts.

the problem with this is that most non-contig related xfails only makes sense within the context of a certain class of ops, so I'm not sure we can detangle these entirely.